File: src/board/BoardV0.go
package board

import (
	"github.com/HuXin0817/dots-and-boxes/src/config"
	"github.com/HuXin0817/dots-and-boxes/src/model"
	"github.com/stretchr/testify/assert"
)

type V0 struct {
	model.Step
	edges [model.MaxEdge]model.Edge
	index [model.MaxEdge]int
}

func NewV0() *V0 {
	b := &V0{}
	for i := range model.MaxEdge {
		b.index[i] = int(i)
		b.edges[i] = i
	}
	return b
}

func (b *V0) Add(e model.Edge) {
	if config.DEBUG {
		assert.True(nil, b.NotContains(e))
	}
	r := b.edges[b.Step]
	ie := b.index[e]
	ir := int(b.Step)
	b.edges[ie], b.edges[ir] = b.edges[ir], b.edges[ie]
	b.index[e], b.index[r] = ir, ie
	b.Step++
}

func (b *V0) Contains(e model.Edge) bool { return b.index[e] < int(b.Step) }

func (b *V0) NotContains(e model.Edge) bool { return b.index[e] >= int(b.Step) }

func (b *V0) EmptyEdges() []model.Edge { return b.edges[b.Step:] }

func (b *V0) MoveRecord() []model.Edge { return b.edges[:b.Step] }

File: src/board/BoardV0.cpp
#include "BoardV0.h"

BoardV0::BoardV0() {
  for (int i = 0; i < Edge::Max; i++) {
    EdgeIndexes.At(i) = i;
    Edges.At(i) = i;
  }
}

void
BoardV0::Add(Edge edge) {
  assert(NotContains(edge));
  int nowStep = NowStep();
  Edge nowEdge = Edges.At(nowStep);
  int edgeIndex = EdgeIndexes.At(edge);
  std::swap(Edges.At(edgeIndex), Edges.At(nowStep));
  EdgeIndexes.At(edge) = nowStep;
  EdgeIndexes.At(nowEdge) = edgeIndex;
  Go();
}

bool
BoardV0::Contains(Edge e) const {
  return EdgeIndexes.At(e) < NowStep();
}

bool
BoardV0::NotContains(Edge e) const {
  return EdgeIndexes.At(e) >= NowStep();
}

[[nodiscard]] Span<const Edge>
BoardV0::EmptyEdges() const {
  return {Edges.begin() + NowStep(), Edges.begin() + Edge::Max};
}

[[nodiscard]] Span<const Edge>
BoardV0::MoveRecord() const {
  return {Edges.begin(), Edges.begin() + NowStep()};
}
